hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,857 @@
+    Src: DmaBuffer<B>,
+    Dst: DmaBuffer<B>,
+{
+    /// Issue a software trigger request to the corresponding channel.
+    /// Note that is not guaranteed that the trigger request will register,
+    /// if a trigger request is already pending for the channel.
+    #[inline(always)]
+    pub fn software_trigger(&mut self, dmac: &mut DmaController) {
+        self.xfer.software_trigger(dmac);
+    }
+    /// Blocking; Wait for the DMA transfer to complete and release all owned
+    /// resources
+    #[inline(always)]
+    pub fn wait(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        self.xfer.wait(dmac)
+    }
+
+    /// Non-blocking; Immediately stop the DMA transfer and release all owned
+    /// resources
+    #[inline(always)]
+    pub fn stop(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        self.xfer.stop(dmac)
+    }
+}
+
+/// Useable beat sizes for DMA transfers
+#[derive(Clone, Copy)]
+pub enum BeatSize {
+    /// Byte = [`u8`](core::u8)
+    Byte = 0x00,
+    /// Half word = [`u16`](core::u16)
+    HalfWord = 0x01,
+    /// Word = [`u32`](core::u32)
+    Word = 0x02,
+}
+
+// Trait enabling taking `*mut` references to references to a single `T`,
+/// references to arrays of `T` or slices of `T`, where `T: DmaBeat`,
+/// as well as required transfer length
+pub trait DmaBuffer<T: DmaBeat> {
+    /// Take a `*mut` reference to the last object
+    /// in the buffer needed by the DMAC. The pointer
+    /// changes depending on whether the address should
+    fn to_dma_ptr(&mut self) -> *mut T;
+    /// Check if the DMA address should be incrementing or not
+    fn incrementing(&self) -> bool;
+}
+
+impl<T: DmaBeat, const N: usize> DmaBuffer<T> for &mut [T; N] {
+    #[inline(always)]
+    fn to_dma_ptr(&mut self) -> *mut T {
+        let ptr = self.as_mut_ptr();
+        if self.incrementing() {
+            unsafe { ptr.add(N) }
+        } else {
+            ptr
+        }
+    }
+
+    #[inline(always)]
+    fn incrementing(&self) -> bool {
+        N > 1
+    }
+}
+
+impl<T: DmaBeat> DmaBuffer<T> for &mut [T] {
+    #[inline(always)]
+    fn to_dma_ptr(&mut self) -> *mut T {
+        let ptr = self.as_mut_ptr();
+        if self.incrementing() {
+            unsafe { ptr.add(self.len()) }
+        } else {
+            ptr
+        }
+    }
+
+    #[inline(always)]
+    fn incrementing(&self) -> bool {
+        self.len() > 1
+    }
+}
+
+impl<T: DmaBeat> DmaBuffer<T> for &mut T {
+    #[inline(always)]
+    fn to_dma_ptr(&mut self) -> *mut T {
+        *self as *mut T
+    }
+
+    #[inline(always)]
+    fn incrementing(&self) -> bool {
+        false
+    }
+}
+
+impl<T: DmaBeat> DmaBuffer<T> for *mut T {
+    #[inline(always)]
+    fn to_dma_ptr(&mut self) -> *mut T {
+        *self
+    }
+
+    fn incrementing(&self) -> bool {
+        false
+    }
+}
+
+/// Convert 8, 16 and 32 bit types
+/// into [`BeatSize`](BeatSize)
+pub trait DmaBeat {
```

Comment:
I would add `Sealed` as a super trait here. You don't want users to implement their own beat sizes.

Reply:
I'm not sure if it would help, but you could also use `num` traits as super traits as well, e.g. `PrimInt`, `AsPrimitive`, etc. That would allow you to work more easily with generic `DmaBeat`s, if that's a problem at all.